[documentation] Patch Release Instructions and Schedule (#501)
ros-planning/moveit.ros.org: documentation/contributing/releases/index.markdown

---
author: tylerjw
layout: page
slug: releases
title: Releases
---

# Release Process

MoveIt has a 6 week cadence for releases.  This pace is a balance between releasing new improvements quickly and preserving the stability of releases.  We are choosing this schedule to reduce the amount of work per release, reduce the time it takes to get an improvement released and make it easier to schedule for the Picknik team.  These releases will include the main MoveIt repo along with any dependencies we control if they have changed (e.g. moveit_msgs).

## Upcoming releases

> The release schedule is avalible via [this Google Calendar](https://calendar.google.com/calendar/embed?src=c_bpueq5jcajuaejr2di0hdr4qgg%40group.calendar.google.com&ctz=America%2FDenver).

Reviewer: rather than href, can you grab the html from the calendar such that it iframes? 
https://support.google.com/calendar/answer/41207?hl=en

markdown supports arbitrary html

Reviewer 2: ```suggestion
The release schedule is shown here:

<iframe src="https://calendar.google.com/calendar/embed?src=c_bpueq5jcajuaejr2di0hdr4qgg%40group.calendar.google.com&ctz=America%2FDenver" width="100%" height="350px" title="MoveIt release calendar"></iframe>
```

### Typical Release Schedule

* Friday: Begin preparing release
* Following Friday: Bloom
* Tuesday After Bloom: Public Announcement

## Permissions

The maintainer handling the release must have write access to both devel repos (e.g. [github.com/ros-planning/moveit](https://github.com/ros-planning/moveit)) as well as release repos (e.g. [github.com/ros-gbp/moveit-release](https://github.com/ros-gbp/moveit-release)). The release repo access is managed via the Github Team [moveit-maintainers](https://github.com/orgs/ros-gbp/teams/moveit-maintainers/members).

## Steps

> 1. Create an issue to track the status of the upcomming release.  This issue should include a list of the changes since the last release so maintainers can determine which ones should be backported into the release branch.  Do this for each repo that will be released with references to the issue in the main repo.  To generate the list of changes use the `review-for-backport` branch. The `reviewed-for-backport` branch exists as a marker in the commit history for `master`, marking the last commit that was considered for backporting. Run `git cherry -v <release-branch> master reviewed-for-backport | grep '^+'` to generate a list of all commits between `reviewed-for-backport` and the current `master` branch that have not been merged in `release-branch` (without modifications). Backports should be pull-requests and reviewed separately, though possibly bundled.

Reviewer 2: ```suggestion
1. Create an issue to track the status of the upcoming release.  This issue should include a list of the changes since the last release so maintainers can determine which ones should be backported into the release branch.  Do this for each repo that will be released with references to the issue in the main repo.  To generate the list of changes use the `review-for-backport` branch. The `reviewed-for-backport` branch exists as a marker in the commit history for `master`, marking the last commit that was considered for backporting. Run `git cherry -v <release-branch> master reviewed-for-backport | grep '^+'` to generate a list of all commits between `reviewed-for-backport` and the current `master` branch that have not been merged in `release-branch` (without modifications). Backports should be pull-requests and reviewed separately, though possibly bundled.
```

   * Example commands:
    Default for Kinetic-Xenial where the `HEAD` at the moveit repo is what we want to release, and Kinetic-Wily (basically the same except for the Ubuntu distro type):

     ```
     _DIR_PRLTEST=/tmp/prerelease_job_kin-xen; mkdir -p $_DIR_PRLTEST && cd $_DIR_PRLTEST

     generate_prerelease_script.py   https://raw.githubusercontent.com/ros-infrastructure/ros_buildfarm_config/production/index.yaml melodic default ubuntu xenial amd64  moveit   --level 0  --output-dir ./

     generate_prerelease_script.py   https://raw.githubusercontent.com/ros-infrastructure/ros_buildfarm_config/production/index.yaml melodic default ubuntu wily amd64  moveit   --level 0  --output-dir ./
     ```
1. Update changelogs. Take advantage of `catkin_generate_changelog` command to populate new logs, then preferably edit them manually to cleanup/combine changelog messages and mark/sort them by type of change (e.g. bugfix, feature, maintenance, documentation). Have a look at previous changelogs to get an idea of the verbosity level and formatting. Example of the whole command set:

   ```

>    cd moveit                              (Top directory of your clooned moveit repo.)

Reviewer 2: ```suggestion
   cd moveit                              (Top directory of your cloned moveit repo.)
```

---
author: tylerjw
layout: page
slug: releases
title: Releases

> redirect_from: "/documentation/contributing/releases.html"

Reviewer: i don't think we need this meta data... there isn't currently a page called releases.html that needs redirection

Author: removed.

---

# Patch Releases

> MoveIt has a 6 week cadence for patch releases.  This pace is a balance between releasing new improvements quickly and preserving the stability of releases.  We are choosing this schedule to reduce the amount of work per release, reduce the time it takes to get an improvement released and make it easier to schedule for the Picknik team.  These releases will include the main MoveIt repo along with any dependencies we control if they have changed (ie. moveit_msgs).

Reviewer 3: guess this should be `e.g.` not `ie`
```suggestion
MoveIt has a 6 week cadence for patch releases.  This pace is a balance between releasing new improvements quickly and preserving the stability of releases.  We are choosing this schedule to reduce the amount of work per release, reduce the time it takes to get an improvement released and make it easier to schedule for the Picknik team.  These releases will include the main MoveIt repo along with any dependencies we control if they have changed (e.g. moveit_msgs).
```

## Upcoming releases

* **Noetic (1.1.1) and Foxy (2.0.1)**

>   * Oct 02, 2020 - Code freeze

Reviewer: Perhaps a more efficient way of tracking these dates is via a dedicated GCal that has recurring events for all three of these dates, every 6 weeks. Advantages:

- Can embed that calendar here
- Can add it to your calendar, and anyone else who is curious (Robert), so this doesn't get forgotten
- Will auto-make dates to infinity...

Reviewer 3: We will likely have to deviate from this at some point anyway, so I don't like documenting it here in markdown.

Author: Moved this into a google calendar.  That will make it easier for us to change in the future and if we don't have to change it we won't have to do it via a pull request here.

  * Oct 23, 2020 - Bloom
  * Oct 27, 2020 - Announce
* **Noetic (1.1.2) and Foxy (2.0.2)**
  * Nov 13, 2020 - Code Freeze
  * Dec 04, 2020 - Bloom
  * Dec 08, 2020 - Announce
* **Noetic (1.1.3) and Foxy (2.0.3)**
  * Jan 01, 2021 - Code Freeze
  * Jan 22, 2021 - Bloom
  * Jan 26, 2021 - Announce
* **Noetic (1.1.4) and Foxy (2.0.4)**
  * Feb 12, 2021 - Code Freeze
  * Mar 05, 2021 - Bloom
  * Mar 09, 2021 - Announce
* **Noetic (1.1.5) and Foxy (2.0.5)**
  * Mar 26, 2021 - Code Freeze
  * Apr 16, 2021 - Bloom
  * Apr 20, 2021 - Announce

### Typical Release Schedule

> * Friday: Code Freeze

Reviewer 4: I would name it feature freeze. Also, I think a 3-week code-freeze is not realistic (too long). Given a 6 or 8 weeks schedule, this leaves us only a few weeks to actually merge new features. We should (try to) coordinate preparing a new release within one week.
Backporting changes and fixing bugs shouldn't be part of the release process but should be done in advance.

Author: I think I'm using the wrong term here.  What I meant by "code freeze" is that development would continue on master but that we would have 3 weeks to backport changes into the release branch but wouldn't backport changes landed after this date.

For Noetic this should be relatively trivial as we are just going to have it track the master branch, but in the future, if they diverge it would be nice to have 3 weeks to finish the backports.

Reviewer: +1 Feature Freeze
+1 for shorter than 3 weeks, 1 week seems enough to me since we're doing it so frequently 

> Backporting changes and fixing bugs shouldn't be part of the release process but should be done in advance.

I think I disagree. If we don't have this period scheduled, seems it won't happen and then will delay the release.

Author: I added the google calendar link and updated this to be one week between when we "begin preparing release" and when we bloom.

  * Backport changes
  * Fix bugs
  * Update CHANGELOG
* Three Fridays Later: Release
  * Bloom
* Tuesday After Bloom: Public Announcement

## Permissions

> Release person must have a write access to both devel repos (e.g. [github.com/ros-planning/moveit](https://github.com/ros-planning/moveit)) as well as release repos (e.g. [github.com/ros-gbp/moveit-release](https://github.com/ros-gbp/moveit-release)). The release repo access is managed via the Github Team [moveit-maintainers](https://github.com/orgs/ros-gbp/teams/moveit-maintainers/members).

Reviewer 3: ```suggestion
The maintainer handling the release must have write access to both devel repos (e.g. [github.com/ros-planning/moveit](https://github.com/ros-planning/moveit)) as well as release repos (e.g. [github.com/ros-gbp/moveit-release](https://github.com/ros-gbp/moveit-release)). The release repo access is managed via the Github Team [moveit-maintainers](https://github.com/orgs/ros-gbp/teams/moveit-maintainers/members).
```
